webrtc-c/canary/src/Config.cpp

```diff
@@ -2,29 +2,78 @@
 
 namespace Canary {
 
-STATUS mustenv(CHAR const* pKey, Config::Value<const CHAR*>* pResult)
+STATUS Config::init(INT32 argc, PCHAR argv[])
+{
+    // TODO: Probably also support command line args to fill the config
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(argv != NULL, STATUS_NULL_ARG);
+    MEMSET(this, 0, SIZEOF(Config));
+
+    if (argc == 2) {
+        DLOGI("Reading configuration from %s\n", argv[1]);
+        CHK_STATUS(initWithJSON(argv[1]));
+    }
+
+    CHK_STATUS(initWithEnvVars());
+
+    // Need to impose a min duration
+    if (duration.value != 0 && duration.value < CANARY_MIN_DURATION) {
+        DLOGW("Canary duration should be at least %u seconds. Overriding with minimal duration.",
+              CANARY_MIN_DURATION / HUNDREDS_OF_NANOS_IN_A_SECOND);
+        duration.value = CANARY_MIN_DURATION;
+    }
+
+    // Need to impose a min iteration duration
+    if (iterationDuration.value < CANARY_MIN_ITERATION_DURATION) {
+        DLOGW("Canary iterations duration should be at least %u seconds. Overriding with minimal iterations duration.",
+              CANARY_MIN_ITERATION_DURATION / HUNDREDS_OF_NANOS_IN_A_SECOND);
+        iterationDuration.value = CANARY_MIN_ITERATION_DURATION;
+    }
+
+CleanUp:
+
+    return retStatus;
+}
+
+BOOL strtobool(const CHAR* value)
+{
+    if (STRCMPI(value, "on") == 0 || STRCMPI(value, "true") == 0) {
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+STATUS mustenv(CHAR const* pKey, Config::Value<Config::String>* pResult)
 {
     STATUS retStatus = STATUS_SUCCESS;
+    const CHAR* value;
 
     CHK(pResult != NULL, STATUS_NULL_ARG);
+    CHK(!pResult->initialized, retStatus);
 
-    CHK_ERR((pResult->value = getenv(pKey)) != NULL, STATUS_INVALID_OPERATION, "%s must be set", pKey);
+    CHK_ERR((value = getenv(pKey)) != NULL, STATUS_INVALID_OPERATION, "%s must be set", pKey);
+    SNPRINTF(pResult->value, ARRAY_SIZE(pResult->value), "%s", value);
```

Comment:
Not sure I understand the reason for SNPRINF? Why not STRNCPY? Much faster

Reply:
I agree that STRNCPY is much faster. But, STRNCPY is also more dangerous and we don't need speed in this code path. 

STRNCPY fails to null terminate if the whole buffer is used for string data (i.e. if the source string length is at least as long as the dest buffer)

For example:

```c
  char raw[] = "hello";
  char buff[2] = "aa";
  
  // buff will not be null terminated
  strncpy(buff, raw, 1);
```

Reply:
Understood. Normally we reserve a single char more space, copy N and then set the last to NULL. You can also check if it "overflown" the buffer looking at the result. Using STRNCPY is really weird IMHO. However, I am OK if you decide to leave SNPRINTF

Reply:
In this case why not use `STRNCAT` which will *always* null terminate?

Reply:
`STRNCAT` might be more error prone as the last parameter is not the length, but it's the remaining space. More details about `STRNCPY` and `STRNCAT` buffer overflow vulnerabilities, https://us-cert.cisa.gov/bsi/articles/knowledge/coding-practices/strncpy-and-strncat.

Reply:
Right I just wanted to present an alternative to STRNCPY/SNPRINTF in case you didn't consider it already.  I have no issues with SNPRINTF.